### PR TITLE
Use llvm-symbolizer's JSON output to provide function start lines

### PR DESCRIPTION
**Pull request by insilications**
When analyzing a perf.data profile converted automatically via perf_to_profile via pprof -raw perf.data, no function start lines (s=0) are present in any of the locations. With https://github.com/google/pprof/commit/813a5fbdbec8a66f7a5aedb876e1b2c3ee0f99ac, this can be easily solved by using the same JSON frame data from llvm-symbolizer to provide StartLine for Function.start_line. This solves https://github.com/google/pprof/issues/823.

### CODE REVIEW

**Review by aalexand — DISMISSED** [review]

**Comment by aalexand** [comment]
A few CI tests failed:

```
++ go test github.com/google/pprof/fuzz github.com/google/pprof/internal/binutils github.com/google/pprof/internal/driver github.com/google/pprof/internal/elfexec github.com/google/pprof/internal/graph github.com/google/pprof/internal/measurement github.com/google/pprof/internal/report github.com/google/pprof/internal/symbolizer github.com/google/pprof/internal/symbolz github.com/google/pprof/profile
ok  	github.com/google/pprof/fuzz	0.003s
--- FAIL: TestObjFile (0.80s)
    --- FAIL: TestObjFile/fixed_load_address (0.67s)
        binutils_test.go:379: SourceLine for main: got [{main /tmp/hello.c 3 0 0}]; want [{main /tmp/hello.c 3 0 3}]
    --- FAIL: TestObjFile/simulated_ASLR_address (0.13s)
        binutils_test.go:379: SourceLine for main: got [{main /tmp/hello.c 3 0 0}]; want [{main /tmp/hello.c 3 0 3}]
FAIL
FAIL	github.com/google/pprof/internal/binutils	0.860s
```

**Comment by insilications** [comment]
Fixed TestPEFile windows test that were failing and were obviously wrong. I have no idea why TestObjFile linux tests were failing in the CI tests. Let's try again please?

**Comment by aalexand** [comment]
Looks like the CI still failed.

**Comment by insilications** [comment]
> Maybe the CI setup needs to install llvm-symbolizer.

Could be... The TestPEFile windows tests were fixed because `ci / test-windows` is not failing anymore. I have no idea why it is failing on Linux. In my Linux system I even added a temporary debug log to double check: `t.Logf("DEBUG SourceLine for main: got %v; want %v\n", gotFrames, wantFrames)`:

```
=== RUN   TestObjFile
=== RUN   TestObjFile/fixed_load_address
    binutils_test.go:378: DEBUG SourceLine for main: got [{main /tmp/hello.c 3 0 3}]; want [{main /tmp/hello.c 3 0 3}]
=== RUN   TestObjFile/simulated_ASLR_address
    binutils_test.go:378: DEBUG SourceLine for main: got [{main /tmp/hello.c 3 0 3}]; want [{main /tmp/hello.c 3 0 3}]
--- PASS: TestObjFile (0.10s)
    --- PASS: TestObjFile/fixed_load_address (0.09s)
    --- PASS: TestObjFile/simulated_ASLR_address (0.01s)
```
    

I wish we could solve this because it would close https://github.com/google/pprof/issues/823

**Comment by aalexand** [comment]
> Could be... The TestPEFile windows tests were fixed because `ci / test-windows` is not failing anymore. I have no idea why it is failing on Linux. In my Linux system I even added a temporary debug log to double check: `t.Logf("DEBUG SourceLine for main: got %v; want %v\n", gotFrames, wantFrames)`:

[binutils.go](https://github.com/google/pprof/blob/main/internal/binutils/binutils.go) falls back to nm or binutils' addr2line if llvm-symbolizer is not available. I think this is what might be happening since perhaps the CI environment doesn't have llvm-symbolizer. I suspect we need to figure out how to install llvm-symbolizer around this line: https://github.com/google/pprof/blob/a0b0bb1d4134f37be44de6c328680339b0fc13ad/.github/workflows/ci.yaml#L149

**Comment by insilications** [comment]
> [binutils.go](https://github.com/google/pprof/blob/main/internal/binutils/binutils.go) falls back to nm or binutils' addr2line if llvm-symbolizer is not available. I think this is what might be happening since perhaps the CI environment doesn't have llvm-symbolizer. I suspect we need to figure out how to install llvm-symbolizer around this line:
> 
> https://github.com/google/pprof/blob/a0b0bb1d4134f37be44de6c328680339b0fc13ad/.github/workflows/ci.yaml#L149

I have pushed changes to `ci.yaml` to attempt to fix it. Lets check if it passes now?

**Comment by insilications** [comment]
I ran the CI workflow on my fork. It is now passing all `test-linux` jobs on `ubuntu-22.04`. But it's failing on `ubuntu-20.04`. My added `Verify llvm-symbolizer installation` step shows that `llvm-symbolizer --version` on `ubuntu-22.04` is:

```
llvm-symbolizer
Ubuntu LLVM version 14.0.0
  
  Optimized build.
  Default target: x86_64-pc-linux-gnu
  Host CPU: znver3
```

But on `ubuntu-20.04` is:

```
LLVM (http://llvm.org/):
  LLVM version 10.0.0
  
  Optimized build.
  Default target: x86_64-pc-linux-gnu
  Host CPU: (unknown)
```

**Comment by insilications** [comment]
Last commit addresses the issue of an outdated `llvm-symbolizer` on Ubuntu 20.04. It is now passing all tests.

**Comment by aalexand** [comment]
Sorry it's taking me a while to review this. I've been staring at the code on and off, I'm just not a fan of how much this complicates the CI configuration with this code to get llvm/clang installed. But I guess there is no better way so I should just bite the bullet and LGTM.

**Comment by insilications** [comment]
> Sorry it's taking me a while to review this. I've been staring at the code on and off, I'm just not a fan of how much this complicates the CI configuration with this code to get llvm/clang installed. But I guess there is no better way so I should just bite the bullet and LGTM.

Not a problem at all. Take the time you need to give it a thorough review.

Yeah, the additional complexity is sad but it is required... Ubuntu 20.04 provides LLVM 10's `llvm-symbolizer`, which is very old and does not produce a JSON data frame with `StartLine`. The alternative would be to add a minimum version check (LLVM 14) before the test that uses `llvm-symbolizer` and skip it: https://github.com/google/pprof/blob/a0b0bb1d4134f37be44de6c328680339b0fc13ad/internal/binutils/binutils_test.go#L336

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -144,6 +144,23 @@ jobs:
...
           google-chrome --version
 
+      - name: Add LLVM repository (Ubuntu 20.04 only)
+        if: matrix.os == 'ubuntu-20.04'
+        run: |
+          wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -
+          sudo add-apt-repository "deb http://apt.llvm.org/focal/ llvm-toolchain-focal-14 main"
+          sudo apt-get update
+
+      - name: Install LLVM and Clang
```
Comment (aalexand) [inline .github/workflows/ci.yaml:154]: nit: I propose to say "Install llvm-symbolizer" to make the why clearer.

**Review by aalexand — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (insilications) [reply .github/workflows/ci.yaml]: Done!

**Review by aalexand — APPROVED** [review]

**Pull request merged**